src/rrd/mod.rs

```diff
@@ -0,0 +1,529 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use core::time;
+use sqlx::{self, Sqlite};
+use std::{io::Write, path::Path};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("invalid window, can't be zero")]
+    InvalidWindow,
+
+    #[error("invalid retention, can't be zero or less than window")]
+    InvalidRetention,
+}
+
+/// Slot provides the functionality to set or overwrite the value of any metric
+/// at a specific timestamp.
+#[async_trait]
+pub trait Slot {
+    /// Counter sets (or overrides) the current stored value for this key,
+    /// with value
+    async fn counter(&mut self, key: &str, value: f64) -> Result<()>;
+    /// Key return the key of the slot which is the window timestamp
+    async fn key(&self) -> Result<i64>;
```

Comment:
this should never fail. hence return i64 not a result
```suggestion
    async fn key(&self) -> i64;
```